### ISSUE TITLE
[KDE Plasma Wayland] Window does not appear as of commit 8c613d85886aae0faec540b2d257621ae568d683

### ISSUE THREAD

**Issue by chenxiaolong** [wez/wezterm#1257]
### What Operating System(s) are you seeing this problem on?

Linux Wayland (KDE Plasma 5.23.1 on Arch Linux, but also reproducible with Plasma 5.22 on Fedora 34)

### WezTerm version

wezterm 20211021-185611-04768fd1

### Did you try the latest nightly build to see if the issue is better (or worse!) than your current version?

Yes, and I updated the version box above to show the version of the nightly that I tried

### Describe the bug

In a KDE Plasma session, when opening wezterm with `enable_wayland = true`, the terminal window does not appear. The application does run and shows up in the running apps list in Plasma's panel though.

I did a git bisect and it seems this issue started happening in commit 8c613d85886aae0faec540b2d257621ae568d683.

The issue does not happen with `enable_wayland = false` (ie. running in XWayland).

### To Reproduce

1. Create a configuration with `enable_wayland = true` and nothing else
2. Launch Wezterm in a KDE Plasma Wayland session
3. The terminal window is not visible

### Configuration

```lua
return {
    enable_wayland = true,
}
```

### Expected Behavior

The terminal window should be visible.

### Logs

The debug overlay cannot be activated because the window is not visible. The command line output is:
```
 2021-10-24T19:32:12.254Z INFO  wezterm_mux_server_impl::local > setting up /run/user/1000/wezterm/gui-sock-281065
 2021-10-24T19:32:12.327Z INFO  wezterm_gui::termwindow        > OpenGL initialized! Mesa Intel(R) HD Graphics 630 (KBL GT2) 4.6 (Compatibility Profile) Mesa 21.2.4 is_context_loss_possible=false wezterm version: 20211021-185611-04768fd1
 2021-10-24T19:32:12.327Z WARN  wezterm_term::terminalstate::performer > Unhandled OSC  12 7
```

**Comment by chenxiaolong** [comment]
I did some debugging and it seems to be caused by: https://github.com/wez/wezterm/blob/539d72a97fda4f33c655ed039af5a7093142a7e9/window/src/os/wayland/window.rs#L306-L315

When calling `window.set_decorate(Decorations::None)` or `window.set_decorate(Decorations::ClientSide)`, the window shows up properly (though without server side decorations, as expected). If I make the same call with `Deocrations::ServerSide` or `Decorations::FollowServer`, the window does not show up. If I omit the call completely, then the window shows up properly.

I'm not familiar enough with Wayland and SCTK to understand why this is happening.

In the meantime, I've worked around the problem by changing the code block to:

```rust
        if decorations == WindowDecorations::NONE {
            window.set_decorate(Decorations::None);
        } else if decorations != WindowDecorations::default() {
            // SCTK/Wayland don't allow more nuance than "decorations are hidden",
            // so if we have a mixture of things, then we need to force our
            // client side decoration rendering.
            window.set_decorate(Decorations::ClientSide);
        }
```

so that `window.set_decorate` is never called when server side decorations should be used.

**Comment by chenxiaolong** [comment]
This appears to be fixed in kwin 5.23.90 (5.24 beta). I suspect it's the same bug that impacted Chromium: 
* https://invent.kde.org/plasma/kwin/-/issues/71
* https://bugs.kde.org/show_bug.cgi?id=444962

which was fixed by:
* https://invent.kde.org/plasma/kwayland-server/-/merge_requests/155
* https://invent.kde.org/plasma/kwin/-/merge_requests/1755
* https://invent.kde.org/plasma/kwin/-/merge_requests/1765
